README.md
# FAAS 
![example workflow name](https://github.com/rknizzle/faas/workflows/Test/badge.svg)

### A simple Functions-as-a-Service(FaaS) platform using Docker containers for the purpose of experimentation/learning

- [Dependencies](#dependencies)
- [Install](#install-cli-tool)
- [Demo](#quick-local-demo)
- [Example](#example)
- [API](#rest-api)
  - [Add Function](#add-function)
  - [Invoke Function](#invoke-function)
- [Roadmap](#roadmap)


# Dependencies
- Only works in Linux and tested on Ubuntu
- Must have Docker daemon installed and running 

# Install CLI tool
NOTE: Currently only Linux is supported due to extra effort required to interact with Docker
containers on Mac OS and Windows because of the Docker virtual machine

The latest Linux binary can be found [here](https://github.com/rknizzle/faas/releases)

# Quick Local Demo:

###  Start the server:
`faas start`

### Create and invoke a function
In another terminal:
```sh
mkdir examplefn
cd examplefn
faas init
faas build
faas invoke examplefn
```

Output: `Response: {"hello":"world"}`

# Example
If you are running locally start the faas server first: `faas start`  
Next, create a new working directory for your function. cd into it and then run faas init.  
After running faas init, open index.js in your code editor to modify your function.  

Here is an example function that takes in an array of numbers are returns the sum:  
```js
// takes an array of numbers and returns the sum
module.exports = (context, cb) => {
  let sum
  for (let num of context.numbers) {
    sum += num
  }

  // respond to the caller with the sum of numbers
  return cb({sum: sum})
}
```
Run `faas build` in your functions working directory to deploy the function to the server.  

Now lets put some input data into a JSON file.  
### input.json:  
```json
{
  "numbers": [6,4,13]
}
```

Now invoke the function and pass it your array of numbers:  
`faas invoke fn -d input.json`  

Output: `Response: {"sum":23}`

# REST API
# Add Function

Adds a new function and returns the url used to invoke it

**URL** : `/functions/`

**Method** : `POST`

**Data constraints**

Provide name of the function and a base64 encoded zip file containing the directory with the
function code inside

```json
{
    "name": "<string>",
    "file": "<base64 encoded zip file containing function code>"
}
```

**Data example** All fields must be sent.

```json
{
    "name": "add",
    "file": "UEsDBBQACAAIAIgEkVEAAAAAAAAAAAAAAAAKAAkARG9ja2VyZmlsZVVUBQABAajaXySOwUrGMBAG7/sUS7yJJNVT8WhboUhNiYiKeAjJIq01Cbut4Nv/lNxnvm8enZ0w5Uj3t3cAV9gx+Z3Ql4JxYQp75n94s+6pHx2aQ9gIB+NLOeExye63rdJUKEVKYSGBzs4fWHz48d90rVfJCbUBcK/PmMovLtU7Jx6OFLf6J/ngQNXVqAGG99m+DNg2bQPd1OMnqrNU3aAS4j9ivYrCr0sAAAD//1BLBwgkCHCtpQAAAMMAAABQSwMEFAAIAAgAiASRUQAAAAAAAAAAAAAAAAcACQBmbi55YW1sVVQFAAEBqNpfyknMSy9NTE+1UsjLT0kFBAAA//9QSwcI+HW5hBQAAAAOAAAAUEsDBBQACAAIAIgEkVEAAAAAAAAAAAAAAAAIAAkAaW5kZXguanNVVAUAAQGo2l80jU3KgzAURedZxcWRwoeZf2DX0C3k57WGvuZJ8oKCuPdSqaMLh8M91uLeFDoTHi0HTZLB8kwBnljW0ViLIFlp03NdyvW0U16aIjp1cDmiVTpxcMzehRdUUEhbyXAoVBvrF10KFfOW2JhG2hYpWjGh/3X+EPyA6Ybd4PoIvt9nYpZ/dKsUjt0xmMN8AgAA//9QSwcI5xQrk5AAAAC+AAAAUEsDBBQACAAIAIgEkVEAAAAAAAAAAAAAAAAMAAkAcGFja2FnZS5qc29uVVQFAAEBqNpfTJAxTwQhEIX7/RUTiquUHImJcVtjYW3paYLwzI3eAplB3Yu5/26AM1ryPZj5HtP3RGSSX2BmMlj9Ug4wFw1+QpRzatzZrd0OGqFBuNRzMuDiuZ84Raz2TQcdF9XM1JYQmQqtfU3YZ9qZO5EsM6VMLSAtCPzKiDtDmw1h5UrOTESnPu0dx68ssY17fOrEf9R9lj+LAwck7UXuH25/dQtSRAqMfyIvOR4vixdFf/7srLs5F6T2C0Wg2pMr667tsJhOPwEAAP//UEsHCGpcPl/GAAAALQEAAFBLAwQUAAgACACIBJFRAAAAAAAAAAAAAAAACQAJAHNlcnZlci5qc1VUBQABAajaX0yQQW7rMAxE9zrF7CwB/nKW2fifoVdwbLpRGlAKKQcpCt+9kOygXgl4GI4eOUbWDHolIVX0EHosQcg2O2qc2SJDSujfQfumlzh9fwyiJMfZQv+lihtnhpT8omT/sv6mka1zxnQd7nGYEBj5SpgXHnOIjDFOtH8x87Had4EnevlbMavVKWq2TRf4Gb+oaWGFHi2E1KH/jx+DQ+vFxiWnJbvKUVKby44r7DrQK+QqNEbOQ2ASDHMmwRw46DXwJ2RhLu9Ru44ryZPEj/eoZEvjaooDFy9fbtBivDizOrMvmKJk9DifzqedbBXoy9H9PWgmtiXVwm5Lre43AAD//1BLBwgrY3WL/AAAALgBAABQSwECFAMUAAgACACIBJFRJAhwraUAAADDAAAACgAJAAAAAAAAAAAA7YEAAAAARG9ja2VyZmlsZVVUBQABAajaX1BLAQIUAxQACAAIAIgEkVH4dbmEFAAAAA4AAAAHAAkAAAAAAAAAAADtgeYAAABmbi55YW1sVVQFAAEBqNpfUEsBAhQDFAAIAAgAiASRUecUK5OQAAAAvgAAAAgACQAAAAAAAAAAAO2BOAEAAGluZGV4LmpzVVQFAAEBqNpfUEsBAhQDFAAIAAgAiASRUWpcPl/GAAAALQEAAAwACQAAAAAAAAAAAO2BBwIAAHBhY2thZ2UuanNvblVUBQABAajaX1BLAQIUAxQACAAIAIgEkVErY3WL/AAAALgBAAAJAAkAAAAAAAAAAADtgRADAABzZXJ2ZXIuanNVVAUAAQGo2l9QSwUGAAAAAAUABQBBAQAATAQAAAAA"
}
```

## Success Response

**Condition** : If the function data is valid.

**Code** : `200 OK`

**Content example**

```json
{
    "invoke": "http://localhost:5555/functions/:name",
}
```

## Error Responses

**Code** : `400 BAD REQUEST`

**Content example**

```json
{
    "message": "Failed to deploy function"
}
```

# Invoke Function

Invokes a function

**URL** : `/functions/:name`

**Method** : `POST`

**Data constraints**

Provide the input data to the function in the request body

```json
{
  "anyInputFields": "anyValues"
}
```

**Data example**

```json
{
    "numbers": [7,33,12,18],
}
```

## Success Response

**Condition** : If the function exists.

**Code** : `200 OK`

**Content example**

```json
{
    "response": "[JSON function result]",
}
```

## Error Responses

**Code** : `400 BAD REQUEST`

**Content example**

```json
{
    "message": "[Error message]"
}
```

# Roadmap:
- [x] Pass input data to the function container and recieve a response to return to the caller
- [x] Run unit tests automatically on push/PR via Github Actions
- [x] Automatically deploy a new binary release on PR merge into master via GitHub Actions
- [ ] Store function name/images and users/apps in a datastore
- [ ] Push deployed images to a docker registry and pull images when invoked
- [ ] Decouple the gateway and runners into separate binaries so faas can be scaled across multiple
  machines
- [ ] Support more programming languages


changes.txt
test
test-release


api/api.go
package api

import (
	"archive/zip"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"github.com/gin-gonic/gin"
	"github.com/rknizzle/faas/manager"
	"io"
	"os"
	"path/filepath"
	"strings"
)

type FnData struct {
	File string `json:"file"`
	Name string `json:"name"`
}

func Start() {
	r := gin.Default()
	m := manager.New()

	r.GET("/ping", ping)

	r.POST("/functions", addFunctionHandler(m))
	r.POST("/functions/:fn", invokeHandler(m))

	// Listen and serve on localhost
	r.Run()
}

func ping(c *gin.Context) {
	c.JSON(200, gin.H{
		"message": "pong",
	})
}

func invokeHandler(m *manager.Manager) gin.HandlerFunc {
	return gin.HandlerFunc(func(c *gin.Context) {
		fn := c.Param("fn")
		fmt.Println("Executing function: " + fn)
		m.RunContainer(fn)

		c.JSON(200, gin.H{
			"success": "true",
		})
	})
}

func addFunctionHandler(m *manager.Manager) gin.HandlerFunc {
	return gin.HandlerFunc(func(c *gin.Context) {

		data, err := fnDataFromReq(c)
		if err != nil {
			c.JSON(400, gin.H{
				"message": err.Error(),
			})
		}

		fmt.Println("Creating function...")

		dir, err := dirFromBase64Data(data.Name, data.File)
		if err != nil {
			c.JSON(400, gin.H{
				"message": err.Error(),
			})
		}

		tag := data.Name
		m.BuildImage(dir, tag)

		// remove temporary directory used to build the image
		os.RemoveAll(dir)

		c.JSON(200, gin.H{
			"invoke": c.Request.Host + "/functions/" + tag,
		})
	})
}

func fnDataFromReq(c *gin.Context) (FnData, error) {
	rawData, err := c.GetRawData()
	if err != nil {
		return FnData{}, err
	}

	var fnData FnData
	err = json.Unmarshal(rawData, &fnData)
	if err != nil {
		return FnData{}, err
	}
	return fnData, nil
}

func dirFromBase64Data(fnName string, base64Data string) (string, error) {
	filename, err := writeDataToZip(fnName, base64Data)
	if err != nil {
		return "", err
	}

	// unzip the file into a directory called the function name
	dir := fnName
	_, err = Unzip(filename, dir)
	if err != nil {
		return "", err
	}

	// remove the zip file now that the directory has been extracted
	err = os.Remove(filename)
	if err != nil {
		return "", err
	}

	return dir, nil
}

func writeDataToZip(fnName string, fnData string) (string, error) {
	filename := fnName + ".zip"

	zipFile, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer zipFile.Close()

	decoder := base64.NewDecoder(base64.StdEncoding, strings.NewReader(fnData))

	// write decoded data to zip file
	io.Copy(zipFile, decoder)
	return filename, nil
}

// Function found at https://golangcode.com/unzip-files-in-go/ (MIT License)
// Unzip a file
func Unzip(src string, dest string) ([]string, error) {

	var filenames []string

	r, err := zip.OpenReader(src)
	if err != nil {
		return filenames, err
	}
	defer r.Close()

	for _, f := range r.File {

		// Store filename/path for returning and using later on
		fpath := filepath.Join(dest, f.Name)

		// Check for ZipSlip. More Info: http://bit.ly/2MsjAWE
		if !strings.HasPrefix(fpath, filepath.Clean(dest)+string(os.PathSeparator)) {
			return filenames, fmt.Errorf("%s: illegal file path", fpath)
		}

		filenames = append(filenames, fpath)

		if f.FileInfo().IsDir() {
			// Make Folder
			os.MkdirAll(fpath, os.ModePerm)
			continue
		}

		// Make File
		if err = os.MkdirAll(filepath.Dir(fpath), os.ModePerm); err != nil {
			return filenames, err
		}

		outFile, err := os.OpenFile(fpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
		if err != nil {
			return filenames, err
		}

		rc, err := f.Open()
		if err != nil {
			return filenames, err
		}

		_, err = io.Copy(outFile, rc)

		// Close the file without defer to close before next iteration of loop
		outFile.Close()
		rc.Close()

		if err != nil {
			return filenames, err
		}
	}
	return filenames, nil
}
